walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,90 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // stores conflicts of all required attendees
+    ArrayList<TimeRange> conflicts = new ArrayList<>();
+    // will store available times 
+    ArrayList<TimeRange> free = new ArrayList<>();
+
+    // find conflicts of required attendees
+    for (Event event : events) {
+      for (String attendee : request.getAttendees()) {
+        if (event.getAttendees().contains(attendee)) {
+          conflicts.add(event.getWhen());
+        }
+      }
+    }
+
+    Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+
+    // handle edge cases 
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return free;
+    }
+    int num_conflicts = conflicts.size(); 
+    if (num_conflicts == 0) {
+      free.add(TimeRange.WHOLE_DAY);
+      return free;
+    }
+
+    int start = 0; int end;
+    TimeRange start_interval;
+    TimeRange end_interval;
+
+    // make events in conflicts disjoint
+    while (start < conflicts.size()) {
+      start_interval = conflicts.get(start);
+      end = start + 1;
+      while (end < conflicts.size()){
+        end_interval = conflicts.get(end);
+        if (start_interval.contains(end_interval)) {
+          conflicts.remove(end);
+        }
+        else if (start_interval.overlaps(end_interval)){
+          start_interval = TimeRange.fromStartEnd(start_interval.start(), end_interval.end(), false);
+          conflicts.set(start, start_interval);
+          conflicts.remove(end);
+        }
+        else {
+          end++;
```

Comment:
You can just break here. Since your events are sorted, if any events overlap then they will be adjacent. So if element n doesn't overlap, then you know for certain elements n+1 (and so on) won't overlap either. This changes this chunk of code from being O(n^2) to being O(n)